[그래] API 테스트/문서자동화 미션 제출합니다

hwanghe159 opened the pull request:
잘부탁드립니다!

hwanghe159 committed:
docs : README.md 작성

hwanghe159 committed:
test : 이번 미션에 관련없는 test들 disabled처리, 나머지 테스트들 pass

hwanghe159 committed:
refactor : MemberController -> AdminMemberController 이름 변경

hwanghe159 committed:
test : 로그인 후 회원정보 조회,수정,삭제에 대한 인수테스트 작성

hwanghe159 committed:
test : 회원정보 인수테스트 상세 작성

hwanghe159 committed:
feat : 본인의 정보를 조회, 수정, 삭제하는 로직 구현

hwanghe159 committed:
chore : 문서자동화를 위한 build.gradle 수정

hwanghe159 committed:
test : 사용자 관리 컨트롤러에 대한 테스트 코드 작성

hwanghe159 committed:
feat : 회원가입 frontend 연결

hwanghe159 committed:
feat : 로그인 frontend 연결, 로컬스토리지에 토큰 저장

hwanghe159 committed:
feat : 마이페이지 연동

hwanghe159 committed:
docs : README 수정

hwanghe159 committed:
feat : 로그인, 정보수정 일부구현. 오류 메시지 응답을 위한 ErrorResponse 생성

hwanghe159 committed:
feat : 회원가입 중복체크 추가, 로그인 전에 이미 로그인 되어있는지 확인로직 추가

hwanghe159 committed:
docs : README 수정

hwanghe159 committed:
feat : 회원정보 수정기능 구현

hwanghe159 committed:
test : MemberService에 대한 테스트 보완

hwanghe159 committed:
test : BearerAuthInterceptor에 대한 테스트 작성

hwanghe159 committed:
docs : README 수정

hwanghe159 committed:
feat : 회원정보 삭제 로직 클라이언트와 연결

hwanghe159 committed:
test : MemberService에 대한 테스트 코드 추가

hwanghe159 committed:
feat : Favorite 도메인, 테이블 생성

hwanghe159 committed:
docs : README 수정

hwanghe159 committed:
refactor : Disabled 제거

hwanghe159 committed:
feat : schema 수정, 검색전 로그인 여부 판별로직 추가

hwanghe159 committed:
feat : 즐겨찾기 추가에 대한 인수테스트, 추가로직 구현

hwanghe159 committed:
feat : 즐겨찾기 template 구현

hwanghe159 committed:
feat : 본인의 즐겨찾기 목록 조회 구현

hwanghe159 committed:
feat : DB조회 실패시 발생하는 커스텀 예외 추가

hwanghe159 committed:
feat : 즐겨찾기 삭제 로직, 인수테스트 추가

hwanghe159 committed:
feat : 즐겨찾기 추가기능 client 연동

hwanghe159 committed:
feat : 즐겨찾기 목록 조회 구현중

hwanghe159 committed:
feat : 없는 데이터 DB조회시 발생하는 예외 추가

hwanghe159 committed:
test : FavoriteController에 대한 테스트코드 작성

hwanghe159 committed:
test : FavoriteService에 대한 테스트 작성

hwanghe159 committed:
feat : 즐겨찾기 목록조회, 삭제로직 프론트 연결

hwanghe159 committed:
refactor : 필요없는 메서드, 주석 제거

hwanghe159 committed:
feat : AdminMemberController에 대한 문서 자동화

hwanghe159 committed:
docs : 즐겨찾기 문서자동화

hwanghe159 committed:
docs : README 수정

hwanghe159 committed:
refactor : 필요없는 라인 제거

dave915 commented on src/main/java/wooteco/subway/config/WebMvcConfig.java:
```diff
@@ -3,35 +3,24 @@
-        registry.addInterceptor(basicAuthInterceptor).addPathPatterns("/me/basic");
-        registry.addInterceptor(sessionInterceptor).addPathPatterns("/me/session");
-        registry.addInterceptor(bearerAuthInterceptor).addPathPatterns("/me/bearer");
+        registry.addInterceptor(bearerAuthInterceptor).addPathPatterns("/me", "/me/**");
```
보통 인증처리 관련 인터셉터를 걸땐 전체를 걸고 인증처리 하지 않을 api만 exclude 해주는 것이 인증 누락될 확률이 낮을 것 같아요 :)

hwanghe159 replied:
그렇군요!! 수정했습니다

dave915 commented on src/main/java/wooteco/subway/domain/member/Member.java:
```diff
@@ -2,13 +2,18 @@
     private String email;
     private String name;
     private String password;
+    private Set<Favorite> favorites = new HashSet<>();
```
Member가 매번 favorites를 갖고있는 것이 맞는가에 대해 의문이 드는데, 그래는 어떻게 생각하시나요?

hwanghe159 replied:
저도 처음에 구현할 때 고민했던 부분이에요!

`Member`가 `favorites`를 갖고 있는게 직관적이고, `Member`가 지워지면 `favorites`들도 지워져야 한다는 생각에 우선 이렇게 구현했었어요.

근데 이 방법은 `favorites`이 당장은 필요없는 로직(개인정보 수정 등)에서도 항상 `favorites`을 갖고 있어야 해서 성능 이슈가 있을 것 같아요.

dave915 replied:
>Member가 favorites를 갖고 있는게 직관적이고, Member가 지워지면 favorites들도 지워져야 한다는 생각에 우선 이렇게 구현했었어요

이부분은 로직적으로 처리 가능하고, DB기능상으로도 처리 가능하기 때문에 조회할때 꼭 함께 조회되어야 하는가를 먼저 고려하면 좋을 것 같아요. 지금은 favorites 하나지만 비슷한 테이블이 많아졌을때를 생각해 보면 될 것같아요 :)

hwanghe159 replied:
함께 조회될 필요가 없다고 판단돼서 Member와 Favorite을 분리했습니다!

dave915 commented on src/main/java/wooteco/subway/web/member/NoSuchFavoriteException.java:
```diff
@@ -0,0 +1,14 @@
+import org.springframework.web.bind.annotation.ResponseStatus;
+
+@ResponseStatus(HttpStatus.NO_CONTENT)
+public class NoSuchFavoriteException extends RuntimeException {
```
NoSuchFavoriteException은 NoSuchFavoriteException 패키지에 모아두는게 좋지 않을까요?

hwanghe159 replied:
exception 패키지 말씀하시는 거죠??
예외 관련 클래스들을 모두 exception 패키지로 모았습니다.

dave915 commented on src/main/java/wooteco/subway/web/ExceptionController.java:
```diff
@@ -0,0 +1,42 @@
+
+    @ExceptionHandler(InvalidAuthenticationException.class)
+    public ResponseEntity<ErrorResponse> InvalidAuthenticationExceptionHandler(InvalidAuthenticationException e) {
+        return ResponseEntity.status(HttpStatus.UNAUTHORIZED).body(new ErrorResponse("비밀번호를 잘못 입력하셨습니다.", 401));
```
log를 남기는 법을 찾아본 뒤 logger.error("error >> {}", e.getMessage(), e); 이런식으로 로그를 남겨보세요!
보통은 개발자가 장애 대응 하기 쉽게 log에 자세한 내용을 남기고, response에는 대략적인 메세지만 전달해줘요~

hwanghe159 replied:
좋은 정보 감사합니다!! 코드에 적용했습니다.

dave915 commented on src/main/java/wooteco/subway/web/ExceptionController.java:
```diff
@@ -0,0 +1,42 @@
+import wooteco.subway.web.member.NoSuchMemberException;
+
+@ControllerAdvice
+public class ExceptionController {
```
예상하지 못한 Exception도 처리 할 수 있게 Exception.class도 처리 할 수 있는 메소드를 만들었으면 좋겠어요 :)

hwanghe159 replied:
추가했습니다!

dave915 commented on src/main/java/wooteco/subway/service/member/MemberService.java:
```diff
@@ -1,40 +1,70 @@
         Member member = memberRepository.findByEmail(param.getEmail()).orElseThrow(RuntimeException::new);
         if (!member.checkPassword(param.getPassword())) {
-            throw new RuntimeException("잘못된 패스워드");
+            throw new InvalidAuthenticationException("잘못된 패스워드");
```
혹시 잘못된 패스워드일 경우 InvalidAuthenticationException 발생하는 테스트가 있나요?

hwanghe159 replied:
그 부분을 놓쳤네요 😭 
잘못된 패스워드에 대한 테스트 코드 추가했습니다.

dave915 commented on src/main/java/wooteco/subway/service/member/MemberService.java:
```diff
@@ -44,8 +74,32 @@ public Member findMemberByEmail(String email) {
+    }
+
+    public Member addFavorite(Long memberId, FavoriteCreateRequest request) {
+        Member member = memberRepository.findById(memberId).orElseThrow(NoSuchMemberException::new);
```
MethodArgumentsResolver에서 이미 db조회를 해서 넣어준 상황이라서 그 Member를 사용해도 되지 않을까요?

hwanghe159 replied:
불필요하게 db조회를 2번하고 있었네요!!
해당하는 부분 모두 수정했습니다.!

dave915 reviewed (changes requested):
그래 안녕하세요! 리뷰가 늦었네요 ㅠㅠ 
몇가지 피드백 남겼으니 확인 후 반영 해 보세요 :)

hwanghe159 committed:
refactor : exception 패키지 생성, 예외관련 클래스 exception 패키지로 이동. ExceptionController에서 GlobalExceptionHandler으로 이름 변경

hwanghe159 committed:
refactor : 피드백 반영

- 인증처리 관련 인터셉터 설정 수정
- 잘못된 비밀번호 요청관련 테스트 작성
- 불필요하게 멤버조회를 중복으로 수행하는 로직 개선

hwanghe159 committed:
refactor : 에러메시지 client 연결, 코드 정렬

dave915 commented on src/main/java/wooteco/subway/exception/GlobalExceptionHandler.java:
```diff
@@ -0,0 +1,53 @@
+    @ExceptionHandler(Exception.class)
+    public ResponseEntity<String> ExceptionHandler(Exception e) {
+        LOGGER.error("error >>> {}", e.getMessage(), e);
+        return ResponseEntity.status(HttpStatus.INTERNAL_SERVER_ERROR).body(SYSTEM_ERROR_MESSAGE);
```
👍

dave915 commented on src/docs/asciidoc/api-guide.adoc:
```diff
@@ -14,9 +14,55 @@ endif::[]
+
+[[resources-members-update]]
+=== 회원 수정
+operation::members/update[snippets='http-request,http-response']
```
path-parameters, request-fields도 함께 써줘서 해당 field에 관한 설명도 함께 제공해주면 좋을 것 같아요 :) 

```suggestion
operation::members/update[snippets='http-request,http-response,path-parameters,request-fields']
```

hwanghe159 replied:
모든 api에 추가했습니다!!

dave915 reviewed (changes requested):
그래! 피드백 반영 확인 했어요 :) 
코멘트 남긴부분에 코멘트 남겼으니 확인 부탁드릴게요 :)

hwanghe159 committed:
docs : api-guide.adoc 더 많은 정보를 나타내도록 수정

hwanghe159 committed:
refactor : Member가 favorite을 가지지 않도록 수정, 필요없는 메서드, 클래스 제거

hwanghe159 committed:
refactor : FAVORITE의 member칼럼명 변경

hwanghe159 committed:
refactor : 즐겨찾기 목록 프론트 연결

dave915 commented on src/main/java/wooteco/subway/service/favorite/FavoriteService.java:
```diff
@@ -0,0 +1,49 @@
+
+    public List<FavoriteResponse> findAllFavoriteResponses(Long memberId) {
+        List<FavoriteResponse> favoriteResponses = new ArrayList<>();
+        List<Favorite> favorites = favoriteRepository.findAllByMemberId(memberId);
+        for (Favorite favorite : favorites) {
+            Station sourceStation = stationRepository.findById(favorite.getSourceStationId()).orElseThrow(NoSuchStationException::new);
+            Station targetStation = stationRepository.findById(favorite.getTargetStationId()).orElseThrow(NoSuchStationException::new);
+            String sourceStationName = sourceStation.getName();
+            String targetStationName = targetStation.getName();
+            favoriteResponses.add(FavoriteResponse.of(favorite, sourceStationName, targetStationName));
+        }
+        return favoriteResponses;
```
favorites를 loop를 돌면서 DB에 요청하면 너무 많이 DB에요청이 가는데, 조회 할 stationId들을 추려서 한번만 DB에 요청 후 결과를 맵핑하는 방식으로 구현해보면 어떨까요?

hwanghe159 replied:
for문안에서 DB요청하던 걸 밖으로 빼서 한번에 조회하도록 수정했습니다!

dave915 reviewed (changes requested):
그래! 피드백 반영한 부분 확인 했어요. 
코멘트 하나 남겼는데, 확인 후 반영 해 보세요 :) 
혹시 궁금한 부분있다면 편하게 코멘트나 DM주세요!

hwanghe159 committed:
refactor : DB요청 횟수 줄임, FavoriteResponses 일급컬렉션 생성

hwanghe159 committed:
refactor : 피드백 반영

- Map을 사용하여 DB요청수 줄임
- FavoriteResponses 일급컬렉션 제거
- Favorites, Stations 일급컬렉션 생성, 테스트 작성

hwanghe159 committed:
refactor : 코드 정렬

dave915 commented on src/main/java/wooteco/subway/service/favorite/FavoriteService.java:
```diff
@@ -21,19 +21,12 @@ public FavoriteService(FavoriteRepository favoriteRepository, StationRepository
-        List<Station> targetStations = stationRepository.findAllById(targetStationIds);
-
-        return FavoriteResponses.of(favorites, sourceStations, targetStations);
+        return favorites.toFavoriteResponses(stations.toNamesMap());
```
구현 잘 해주셨네요! 💯💯💯

dave915 reviewed (approved):
그래! 피드백 반영 확인 했어요 💯 
이만 머지할게요 :) 고생하셨습니다!
